isofit/radiative_transfer/radiative_transfer.py

```diff
@@ -297,64 +293,61 @@ def drdn_dRT(
-
-        elif self.topography_model:
-            # jac w.r.t. topoflux correct radiance
-            I = self.solar_irr / np.pi
-            t_dir_down = r["transm_down_dir"]
-            t_dif_down = r["transm_down_dif"]
-            if geom.cos_i is None:
-                cos_i = self.coszen
-            else:
-                cos_i = geom.cos_i
-            t_total_up = r["transm_up_dif"] + r["transm_up_dir"]
-            t_total_down = t_dir_down + t_dif_down
-            s_alb = r["sphalb"]
+        # get needed rt quantities from LUT
+        r = self.get_shared_rtm_quantities(x_RT, geom)
 
-            a = t_total_up * (I * cos_i * t_dir_down + I * self.coszen * t_dif_down)
-            drdn_drfl = a / (1 - s_alb * rfl) ** 2
+        # atmospheric spherical albedo
+        s_alb = r["sphalb"]
 
-        elif self.glint_model:
-            L_down_transmitted = self.get_L_down_transmitted(x_RT, geom)
+        # direct and diffuse downward fluxes on the sun-to-surface path
+        # note: currently, E_down_dir comes scaled by the TOA solar zenith angle,
+        # thus, unscaling and rescaling by local solar zenith angle required
+        # to account for surface slope and aspect
+        E_down_dir, E_down_dif = self.get_E_down(x_RT, geom)
+        E_down_dir = E_down_dir / self.coszen * cos_i
 
-            E_dd = (self.solar_irr * self.coszen) / np.pi * r["t_down_dir"]
-            E_ds = (self.solar_irr * self.coszen) / np.pi * r["t_down_dif"]
-            E_d = E_dd + E_ds
-            L_sky = x_surface[-2] * E_dd + x_surface[-1] * E_ds
+        # including glint for water surfaces
+        if self.glint_model:
+            E_down_tot = E_down_dir + E_down_dif
+            L_sky = x_surface[-2] * E_down_dir + x_surface[-1] * E_down_dif
 
             rho_ls = 0.02  # fresnel reflectance factor (approx. 0.02 for nadir view)
-            glint = rho_ls * (L_sky / E_d)
-
-            drho_scaled_for_multiscattering_drfl = (
-                1.0 / (1 - r["sphalb"] * (rfl + glint)) ** 2
-            )
-            drdn_drfl = L_down_transmitted * drho_scaled_for_multiscattering_drfl
-
+            glint = rho_ls * (L_sky / E_down_tot)
         else:
-            L_down_transmitted = self.get_L_down_transmitted(x_RT, geom)
+            glint = np.zeros(rfl.shape)
 
-            # The reflected downwelling light is:
-            # L_down_transmitted * rfl / (1.0 - r['sphalb'] * rfl), or
-            # L_down_transmitted * rho_scaled_for_multiscattering
-            # This term is the derivative of rho_scaled_for_multiscattering
-            drho_scaled_for_multiscattering_drfl = 1.0 / (1 - r["sphalb"] * rfl) ** 2
+        # adjacency effects
+        bg = geom.bg_rfl if geom.bg_rfl is not None else rfl + glint
 
-            drdn_drfl = L_down_transmitted * drho_scaled_for_multiscattering_drfl
+        # K surface reflectance
+        drdn_drfl = (E_down_dir + E_down_dif) / (1.0 - s_alb * bg) * r["transm_up_dir"]
```

Comment:
@pgbrodrick @davidraythompson This derivative is tricky. It's basically only valid if we incorporate a background reflectance term that's different from our target reflectance. If we don't have that at hand, we multiply the spherical albedo by the target reflectance, which would change the denominator of the derivative back to `(1 - r["sphalb"] * rfl) ** 2`. Do we assume to run with background reflectance by default in the future or better implement an if-clause here to capture the 'classic' case?

Reply:
@pgbrodrick @davidraythompson If you have a minute, please let me know your opinion here.

Reply:
I think we're okay with this as is, because the bg term in 343 handles the if/else clause, and all assumptions trickle out from that.